code/modules/spells/spell_types/shapeshift.dm

```diff
@@ -46,10 +46,33 @@
 
 		var/obj/shapeshift_holder/S = locate() in M
 		if(S)
-			Restore(M)
+			M = Restore(M)
 		else
-			Shapeshift(M)
-
+			M = Shapeshift(M)
+		if(M.movement_type & (VENTCRAWLING))
+			if(!M.ventcrawler) //you're shapeshifting into something that can't fit into a vent
+				var/obj/machinery/atmospherics/pipeyoudiein = M.loc
+				var/datum/pipeline/ourpipeline
+				var/pipenets = pipeyoudiein.returnPipenets()
+				if(islist(pipenets))
+					ourpipeline = pipenets[1]
+				else
+					ourpipeline = pipenets
+
+				to_chat(M, "<span class='userdanger'>Casting [src] inside of [pipeyoudiein] quickly turns you into a bloody mush!</span>")
+				for(var/obj/machinery/atmospherics/components/unary/possiblevent in range(10, get_turf(M)))
+					if(possiblevent.parents.len && possiblevent.parents[1] == ourpipeline)
+						if(isalien(M))
+							new /obj/effect/gibspawner/xeno(get_turf(possiblevent))
+						else if(ishuman(M))
+							new /obj/effect/gibspawner/human(get_turf(possiblevent))
+						else
+							new /obj/effect/gibspawner/generic(get_turf(possiblevent))
```

Comment:
you should only have to check for the gibtype once (pre-loop), then the loop just poops out the selected gibtype.

Reply:
done!